src/utils/hotkeys.py

```diff
@@ -0,0 +1,354 @@
+    Skill4 = 'Skill4',
+    Skill5 = 'Skill5',
+    Skill6 = 'Skill6',
+    Skill7 = 'Skill7',
+    Skill8 = 'Skill8',
+    Skill9 = 'Skill9',
+    Skill10 = 'Skill10',
+    Skill11 = 'Skill11',
+    Skill12 = 'Skill12',
+    Skill13 = 'Skill13',
+    Skill14 = 'Skill14',
+    Skill15 = 'Skill15',
+    Skill16 = 'Skill16',
+    SelectPreviousSkill = 'SelectPreviousSkill',
+    SelectNextSkill = 'SelectNextSkill',
+    ShowBelt = 'ShowBelt',
+    UseBelt1 = 'UseBelt1',
+    UseBelt2 = 'UseBelt2',
+    UseBelt3 = 'UseBelt3',
+    UseBelt4 = 'UseBelt4',
+    SwapWeapons = 'SwapWeapons',
+    Chat = 'Chat',
+    Run = 'Run',
+    ToggleRunWalk = 'ToggleRunWalk',
+    StandStill = 'StandStill',
+    ForceMove = 'ForceMove',
+    ShowItems = 'ShowItems',
+    ShowPortraits = 'ShowPortraits',
+    Automap = 'Automap',
+    CenterAutomap = 'CenterAutomap',
+    FadeAutomap = 'FadeAutomap',
+    PartyOnAutomap = 'PartyOnAutomap',
+    NamesOnAutomap = 'NamesOnAutomap',
+    ToggleMiniMap = 'ToggleMiniMap',
+    SayHelp = 'SayHelp',
+    SayFollowMe = 'SayFollowMe',
+    SayThisIsForYou = 'SayThisIsForYou',
+    SayThanks = 'SayThanks',
+    SaySorry = 'SaySorry',
+    SayBye = 'SayBye',
+    SayNowYouDie = 'SayNowYouDie',
+    SayRetreat = 'SayRetreat',
+    ScreenShot = 'ScreenShot',
+    ClearScreen = 'ClearScreen',
+    ClearMessages = 'ClearMessages',
+    Zoom = 'Zoom',
+    LegacyToggle = 'LegacyToggle',
+    OpenMenu = 'OpenMenu(Esc)',
+
+def discover_hotkey_mappings(saved_games_folder, key_name):
+    global d2r_keymap
+    global _default_left_skill_key_map, _swap_left_skill_key_map, left_skill_key_map, left_skill
+    global _default_right_skill_key_map, _swap_right_skill_key_map, right_skill_key_map, right_skill
+    Logger.debug(f"Detecting hotkeys for {key_name}")
+    templates = template_finder.get_cached_templates_in_dir('assets\\templates\\ui\\skills')
+    d2r_keymap = _parse_key_file(saved_games_folder, key_name)
+    key_list = []
+    for i in range(1, 17):
+        hotkey_name = HotkeyName(f'Skill{i}')
+        if hotkey_name in d2r_keymap:
+            key_list.append(d2r_keymap[hotkey_name])
+    found_keys = []
+    img = grab()
+    starting_left_skill = get_selected_skill(templates, img, Config().ui_roi["skill_left"])
+    starting_right_skill = get_selected_skill(templates, img, Config().ui_roi["skill_right"])
+    _find_keymapping(
+        templates,
+        key_list,
+        starting_left_skill,
+        starting_right_skill,
+        found_keys,
+        _default_left_skill_key_map,
+        _default_right_skill_key_map)
+    for key in found_keys:
+        if key in key_list:
+            key_list.remove(key)
+    _find_keymapping(
+        templates,
+        key_list,
+        starting_left_skill,
+        starting_right_skill,
+        found_keys,
+        _default_left_skill_key_map,
+        _default_right_skill_key_map)
+    for key in found_keys:
+        if key in key_list:
+            key_list.remove(key)
```

Comment:
Why we need to do this twice? Also if we have to, use a for loop.

Reply:
We do this twice in order to try to confidently deduce that the key that was pressed has actually changed the left / right skill slot. This has been abstracted into a _deduce_hotkeys helper fn